amused.1
.\" Copyright (c) 2022, 2023, 2024, 2025, 2026 Omar Polo <op@omarpolo.com>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd January 11, 2026
.Dt AMUSED 1
.Os
.Sh NAME
.Nm amused
.Nd music player
.Sh SYNOPSIS
.Nm
.Op Fl dv
.Op Fl s Ar socket
.Oo
.Ar command
.Op Ar argument ...
.Oc
.Sh DESCRIPTION
.Nm
is a simple music player.
It manages a single queue of music files.
At the end of every track,
.Nm
proceeds to the next one in the queue, or with the first one if
it was the last one or if the track wasn't part of the queue.
.Pp
.Nm
automatically skips and removes from the playing queue non-regular files,
or ones with a non-recognized audio format.
.Pp
The following options are available:
.Bl -tag -width Ds
.It Fl d
Do not daemonize:
.Nm
will run in the foreground and log to standard error.
It's ignored if any commands are given on the command line, or if the
server is already running.
.It Fl s Ar socket
Use
.Ar socket
instead of the default
.Pa /tmp/amused-UID
to communicate with the daemon.
.It Fl v
Produce more verbose output.
.It Ar command Op Ar argument ...
Specify the command to run.
If no commands are specified,
.Ic status
is assumed.
Commands may be abbreviated to a unique prefix, for example
.Sq rep
can be given instead of
.Ic repeat .
.El
.Pp
The following commands are available:
.Bl -tag -width Ds
.It Cm add Ar
Enqueue the given files at the end of the playing queue.
.It Cm consume Op Cm on Ns | Ns Cm off
Enable or disable the consume mode.
When consume mode is enabled the tracks are removed from the playing queue
once played completely.
Without arguments toggle the current status.
.It Cm flush
Erase the playlist.
.It Cm jump Ar substring
Play the first song in the playing queue that contains the given
case-insensitive
.Ar substring .
.It Cm load Op Ar file
Load a playlist from
.Ar file
or standard input.
A playlist is a list of paths to music files given one per line
and optionally prefixed by
.Sq > \&
or two spaces.
If the list was generated by
.Nm
.Ic show Fl p ,
restore also the position in the playing queue.
Otherwise, if already playing something, try to match the currently
played song in the new queue.
.It Cm monitor Op Ar events
Stop indefinitely and print when an event in the comma-separated list
of
.Ar events
happened, or all if not given.
The available
.Ar events
are:
.Pp
.Bl -tag -compact -width Ds
.It add
Added a song to the playlist.
.It jump
Jumped to a different song in the playlist.
.It load
Loaded a new playlist.
.It mode
Mode changed
.Pq repeat or consume .
.It next
Advanced to next song.
.It pause
Paused.
.It play
Playing or resuming a song.
.It previous
Gone back to previous song.
.It seek
Seeked in the current track.
.It stop
Stopped.
.El
.It Cm next
Play the next song.
.It Cm pause
Pause the playback.
.It Cm play
Start or resume the playback.
.It Cm previous
Play the previous song.
.It Cm repeat one Ns | Ns Cm all Op Cm on Ns | Ns Cm off
Enable or disable the automatic repetition of the current track
.Pq Cm one
or of the whole playing queue
.Pq Cm all .
Without arguments toggle the given repeat mode.
.It Cm restart
Play the current song from the beginning.
It's a shorthand for
.Nm
.Cm seek
0.
.It Cm seek Oo +- Oc Ns Ar time Ns Op %
Seek by the specified amount of
.Ar time
into the current song.
If the argument is prefixed by either
.Sq +
or
.Sq -
then the seek is relative to the current position, otherwise is
absolute.
With a trailing
.Sq % ,
the
.Ar time
is interpreted as a percentage of the length of the track.
.Pp
The
.Ar time
can also be expressed as
.Oo Oo Ar hours : Oc Ns Ar minutes : Oc Ns Ar seconds ,
but in this case no trailing
.Sq %
is allowed.
.It Cm show Op Fl p
Show the playlist.
With
.Fl p
it prints a
.Dq pretty
list with the current playing song prefixed by
.Sq > \& .
.It Cm shuffle Op Fl a
Shuffle the playing queue after the currently playing track until the
end.
If
.Fl a
is given, all the tracks in the playing queue are shuffled, with
the currently playing one moved at the top.
.It Cm status Op Fl f Ar format
Print playback status and current song.
The
.Ar format
is a comma-separated list of words that select the information to
print, in order.
It defaults to
.Ev AMUSED_STATUS_FORMAT
or to
.Dq status,time:oneline,info,mode:oneline
if not defined.
The formats available are:
.Pp
.Bl -tag -compact -width time:percentage
.It info
Information about the current track.
.It path
Path of the current song
.It mode:oneline
Mode status in a single line.
.It mode
Repeat all, one and consume, one per line.
.It status
Playback status by the path to the current song.
.It time:oneline
Position and duration in a single line.
.It time:percentage
Percentage of the current position.
.It time:raw
Current position and duration in seconds.
.It time
Current position and duration in a human-readable format.
.El
.It Cm stop
Stop the playback.
.It Cm toggle
Play/pause the playback.
.El
.Sh ENVIRONMENT
.Bl -tag -width AMUSED_STATUS_FORMAT
.It Ev AMUSED_STATUS_FORMAT
The default format used by
.Nm
.Cm status .
.It Ev TMPDIR
Path to the directory where the control socket is created.
Defaults to
.Pa /tmp .
.El
.Sh FILES
.Bl -tag -width "/tmp/amused-UID" -compact
.It Pa /tmp/amused-UID
.Ux Ns -domain
socket used for communication with the daemon.
.El
.Sh EXAMPLES
Load every file under the current directory recursively:
.Bd -literal -offset indent
$ find . | amused load
.Ed
.Pp
Enqueue all mp3 files in the current directory:
.Bd -literal -offset indent
$ amused add *.mp3
.Ed
.Pp
Recursively add all opus files:
.Bd -literal -offset indent
$ find . -type f -iname \\*.opus -exec amused add {} +
.Ed
.Pp
Save the state of the player to the file
.Pa amused.dump :
.Bd -literal -offset indent
$ amused show -p > amused.dump
.Ed
.Pp
Load a previous state:
.Bd -literal -offset indent
$ amused load < amused.dump
.Ed
.Pp
Remove duplicates:
.Bd -literal -offset indent
$ amused show | sort | uniq | amused load
.Ed
.Pp
Select a song with
.Xr fzf 1
.Bd -literal -offset indent
$ amused jump "$(amused show | fzf +s)"
.Ed
.Sh SEE ALSO
.Xr amused-mpris2 1 ,
.Xr amused-web 1
.Sh AUTHORS
.An -nosplit
The
.Nm
utility was written by
.An Omar Polo Aq Mt op@omarpolo.com .


CHANGES
* amused 0.19; 2025-05-17
- add compat for reallocarray; should fix build on macOS
- sync all the decoders internal buffer size, and bump it
- use substrings instead of regexp for `amused jump'

* amused 0.18; 2025-02-23
- convert to new imsg APIs and updated bundled imsg
- small tweak to amused.1
- amused-web: add a dark theme

* amused 0.17; 2025-02-15
- sync'ed subcommand usages with the manpage
- renamed the `prev' subcommand to `previous' for consistency
- renamed the `prev' event to `previous' as well
- improvements to the manpage
- add `shuffle' subcommand

* amused 0.16; 2024-09-01
- unbreak the build on various OSes
- add an oss audio backend
- plug a memory leak in amused-web
- add MPRIS2 support via a separate executable amused-mpris2
- amused-web now daemonizes by default; added -d to stay in the foreground
- fixed a bug in ao and oss backends that made amused advance at the wrong rate

* amused 0.15; 2024-05-26
- fixed build of amused-web on linux
- add songmeta: a song metadata extractor; not build by default
- reworked the portable layer; libmd now it's optional on linux and macos
- amused-web now it's built by default; ./configure --without-web to disable
- added ./configure --with-Werror to build with -Werror

* amused 0.14; 2024-01-21
- amused-web: fix typo in error message
- add a libao audio backend
- add an audio backend for android (using oboe)
- add contrib/amused-termux-notification (for android)
- honour $TMPDIR if set
- update bundled imsg
- switched to the new imsg APIs

* amused 0.13; 2023-09-09
- few configure enhancements: allow CC, CFLAGS and LDADD_LIB_* as args too
- fix the compat test for getdtablecount(3)
- drop the dependency on libevent
- add amused-web to control amused remotely (optional; needs libmd on linux)
- contrib/amusing: use fancy unicode characters for the buttons
- unbreak percentage relative seeking
- contrib/amused-monitor: fix handling of non-ASCII characters (finally!)
- update bundled imsg
- add missing HAVE_IMSG in config.h (from tb@; thanks!)

* amused 0.12; 2023-05-17
- few ALSA backend bug fixes
- set up the onmove callback on ALSA too
- fix build with BACKEND=alsa; from ben, thanks!

* amused 0.11; 2023-05-02
- added support for ALSA
- retry when sio_revents fails with EAGAIN
- log errno too when sio_revents fails
- added contrib/amusing: a GUI written in Tcl/Tk
- amused-monitor: use new extended events info
- enriched `mode' and `seek' events reporting in `amused monitor'
- various minor improvements to amused.1
- amused-monitor: fix error for tracks hours long
- amused-monitor: add pledge
- amused-monitor: fix rewind
- amused-monitor: fix off by one
- fix sndio detection

* amused 0.10; 2022-07-13
- changed the default output of 'amused status'
- print the status after 'amused consume' or 'amused repeat'
- fixed 'amused load' with an empty playlist
- the on|off argument to 'amused consume/repeat' is now optional
- added a consume mode
- pledge earlier
- rework (and document) the events reported by "amused monitor" (some were added/removed)
- added contrib/amused-monitor: ncurses tui written in perl (not installed by default)
- added 'amused status -f <fmt>' to customize the output format
- clarify how non-regular or non-supported files in the playlist are treated
- handle fdopen failures gracefully
- use (a modified version of) kristaps' oconfigure; amused is portable now!
- added ability to seek
- keep track of current position and total duration of the track
- don't get stuck if the server fails to start
- improved error reporting when failing to play a song

* amused 0.9; 2022-06-12
- reset the path buffer; fixes the "malformed data" error during `load'

* amused 0.8; 2022-06-12
- try to recover from mp3 decoding errors
- plug a memory leak in player_dispatch
- don't use realpath(3) in `amused add'
- use a stricter pledge for `amused add'
- use canonpath (from kern_pledge.c) for paths handling
- non-blocking usage of sndio: avoids artificial gaps between tracks
- use sio_flush to stop the playback
- automatically skip all non-regular files, not only directories

* amused 0.7; 2022-05-19
- amused monitor optionally accepts a list of events to report
- consume all enqueued messages before calling imsg_read

* amused 0.6; 2022-05-10
- manpage tweaks
- fix file descriptor leak on player_playnext failure
- added a simple filetype detector instead of relying on file extensions

* amused 0.5; 2022-03-26
- unbreak opus and ogg vorbis
- handle flacs with bps != 16 and/or channels != 2
- speed up `load' a lot (especially over NFS)

* amused 0.4; 2022-03-14
- second part of the "unbreak flac" due to carelessly refactoring

* amused 0.3; 2022-03-13
- unbreak flac: add back missing counter reset

* amused 0.2; 2022-03-10
- keep the current song if `load' input was generated by show -p
- update the current song after `load' if the player was stopped
- follow tmux behavior wrt how the daemon is automatically started
- changed the output of `amused status' to simplify parsing
- correctly handle arguments for subcommands
- correctly pledge the player process with "stdio recvfd audio"
- manpage improvements

* amused 0.1; 2022-02-24
- first public version (and probably last too, the thing is finished)


amused.c
/*
 * Copyright (c) 2022 Omar Polo <op@omarpolo.com>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/queue.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/wait.h>

#include <errno.h>
#include <fcntl.h>
#include <imsg.h>
#include <limits.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <syslog.h>
#include <unistd.h>

#include "amused.h"
#include "control.h"
#include "ev.h"
#include "log.h"
#include "player.h"
#include "playlist.h"
#include "xmalloc.h"

char		*csock = NULL;
int		 debug;
int		 verbose;
struct imsgev	*iev_player;

const char	*argv0;
pid_t		 player_pid;

struct player_status current_status;

enum amused_process {
	PROC_MAIN,
	PROC_PLAYER,
};

static __dead void
main_shutdown(void)
{
	pid_t	pid;
	int	status;

	/* close pipes. */
	close(iev_player->imsgbuf.fd);
	imsgbuf_clear(&iev_player->imsgbuf);
	free(iev_player);

	log_debug("waiting for children to terminate");
	do {
		pid = wait(&status);
		if (pid == -1) {
			if (errno != EINTR && errno != ECHILD)
				fatal("wait");
		} else if (WIFSIGNALED(status))
			log_warnx("player terminated; signal %d",
			    WTERMSIG(status));
	} while (pid != -1 || (pid == -1 && errno == EINTR));

	log_info("terminating");
	exit(0);
}

static void
main_sig_handler(int sig, int event, void *arg)
{
	/*
	 * Normal signal handler rules don't apply because ev.c
	 * decouples for us.
	 */

	switch (sig) {
	case SIGTERM:
	case SIGINT:
		main_shutdown();
		break;
	default:
		fatalx("unexpected signal %d", sig);
	}
}

static void
main_dispatch_player(int sig, int event, void *d)
{
	char		*errstr;
	struct imsgev	*iev = d;
	struct imsgbuf	*imsgbuf = &iev->imsgbuf;
	struct imsg	 imsg;
	struct ibuf	 ibuf;
	struct player_status ps;
	size_t		 datalen;
	ssize_t		 n;
	int		 shut = 0;

	if (event & EV_READ) {
		if ((n = imsgbuf_read(imsgbuf)) == -1)
			fatal("imsg_read error");
		if (n == 0)	/* Connection closed */
			shut = 1;
	}
	if (event & EV_WRITE) {
		if (imsgbuf_write(imsgbuf) == -1)
			fatal("imsgbuf_write");
	}

	for (;;) {
		if ((n = imsg_get(imsgbuf, &imsg)) == -1)
			fatal("imsg_get");
		if (n == 0)	/* No more messages. */
			break;

		switch (imsg_get_type(&imsg)) {
		case IMSG_META:
			if (imsg_get_data(&imsg, &ps, sizeof(ps)) == -1)
				fatalx("IMSG_META: got wrong size");
			if (current_status.position != ps.position)
				control_notify(IMSG_CTL_SEEK);
			current_status.duration = ps.duration;
			current_status.position = ps.position;
			if (current_status.duration < 0)
				current_status.duration = -1;
			if (current_status.position < 0)
				current_status.position = -1;
			current_status.info.bits = ps.info.bits;
			current_status.info.rate = ps.info.rate;
			current_status.info.chan = ps.info.chan;
			break;
		case IMSG_ERR:
			if (imsg_get_ibuf(&imsg, &ibuf) == -1 ||
			    (datalen = ibuf_size(&ibuf)) == 0)
				errstr = "unknown error";
			else {
				errstr = ibuf_data(&ibuf);
				errstr[datalen-1] = '\0';
			}
			log_warnx("%s; skipping %s", errstr, current_song);
			playlist_dropcurrent();
			main_playlist_advance();
			if (play_state == STATE_PLAYING)
				control_notify(IMSG_CTL_NEXT);
			else
				control_notify(IMSG_CTL_STOP);
			break;
		case IMSG_EOF:
			if (repeat_one && main_play_song(current_song))
				break;
			else if (repeat_one || consume)
				playlist_dropcurrent();
			main_playlist_advance();
			if (play_state == STATE_PLAYING)
				control_notify(IMSG_CTL_NEXT);
			else
				control_notify(IMSG_CTL_STOP);
			break;
		default:
			log_debug("%s: error handling imsg %d", __func__,
			    imsg_get_type(&imsg));
			break;
		}
		imsg_free(&imsg);
	}

	if (shut)
		ev_break();
	else
		imsg_event_add(iev);
}

static pid_t
start_child(enum amused_process proc, int fd)
{
	const char	*argv[7];
	int		 argc = 0;
	pid_t		 pid;

	if (fd == -1 && debug)
		goto exec;

	switch (pid = fork()) {
	case -1:
		fatal("cannot fork");
	case 0:
		break;
	default:
		close(fd);
		return pid;
	}

	if (fd != 3) {
		if (fd != -1 && dup2(fd, 3) == -1)
			fatal("cannot setup imsg fd");
	} else if (fcntl(F_SETFD, 0) == -1)
		fatal("cannot setup imsg fd");

exec:
	argv[argc++] = argv0;

	switch (proc) {
	case PROC_MAIN:
		argv[argc++] = "-s";
		argv[argc++] = csock;
		argv[argc++] = "-Tm";
		break;
	case PROC_PLAYER:
		argv[argc++] = "-Tp";
		break;
	}

	if (debug)
		argv[argc++] = "-d";
	if (verbose)
		argv[argc++] = "-v";
	argv[argc++] = NULL;

	/* obnoxious casts */
	execvp(argv0, (char *const *)argv);
	fatal("execvp %s", argv0);
}

/* daemon main routine */
static __dead void
amused_main(void)
{
	int	 pipe_main2player[2];
	int	 control_fd, flags;

	log_init(debug, LOG_DAEMON);
	log_setverbose(verbose);
	log_procinit("main");

	if (!debug && daemon(1, 0) == -1)
		fatal("daemon");

	if (socketpair(AF_UNIX, SOCK_STREAM, PF_UNSPEC, pipe_main2player) == -1)
		fatal("socketpair");

	if ((flags = fcntl(pipe_main2player[0], F_GETFL)) == -1 ||
	    fcntl(pipe_main2player[0], F_SETFL, flags | O_NONBLOCK) == -1 ||
	    (flags = fcntl(pipe_main2player[1], F_GETFL)) == -1 ||
	    fcntl(pipe_main2player[1], F_SETFL, flags | O_NONBLOCK) == -1)
		fatal("fcntl(O_NONBLOCK)");

	if (fcntl(pipe_main2player[0], F_SETFD, FD_CLOEXEC) == -1 ||
	    fcntl(pipe_main2player[1], F_SETFD, FD_CLOEXEC) == -1)
		fatal("fcntl(CLOEXEC)");

	player_pid = start_child(PROC_PLAYER, pipe_main2player[1]);

	if (ev_init() == -1)
		fatal("ev_init");

	signal(SIGHUP, SIG_IGN);
	signal(SIGCHLD, SIG_IGN);
	signal(SIGPIPE, SIG_IGN);

	ev_signal(SIGINT, main_sig_handler, NULL);
	ev_signal(SIGTERM, main_sig_handler, NULL);

	iev_player = xmalloc(sizeof(*iev_player));
	if (imsgbuf_init(&iev_player->imsgbuf, pipe_main2player[0]) == -1)
		fatal("imsgbuf_init");
	imsgbuf_allow_fdpass(&iev_player->imsgbuf);
	iev_player->handler = main_dispatch_player;
	iev_player->events = EV_READ;
	ev_add(iev_player->imsgbuf.fd, iev_player->events,
	    iev_player->handler, iev_player);

	if ((control_fd = control_init(csock)) == -1)
		fatal("control socket setup failed %s", csock);
	control_listen(control_fd);

	if (pledge("stdio rpath unix sendfd", NULL) == -1)
		fatal("pledge");

	log_info("startup");
	ev_loop();
	main_shutdown();
}

int
main(int argc, char **argv)
{
	int ch, proc = -1;

	log_init(1, LOG_DAEMON);	/* Log to stderr until daemonized */
	log_setverbose(1);

	argv0 = argv[0];
	if (argv0 == NULL)
		argv0 = "amused";

	while ((ch = getopt(argc, argv, "ds:T:v")) != -1) {
		switch (ch) {
		case 'd':
			debug = 1;
			break;
		case 's':
			free(csock);
			csock = xstrdup(optarg);
			break;
		case 'T':
			switch (*optarg) {
			case 'm':
				proc = PROC_MAIN;
				break;
			case 'p':
				proc = PROC_PLAYER;
				break;
			default:
				usage();
			}
			break;
		case 'v':
			verbose++;
			break;
		default:
			usage();
		}
	}
	argv += optind;
	argc -= optind;

	if (proc == PROC_MAIN)
		amused_main();
	if (proc == PROC_PLAYER)
		exit(player(debug, verbose));

	if (csock == NULL) {
		const char *tmpdir;

		if ((tmpdir = getenv("TMPDIR")) == NULL)
			tmpdir = "/tmp";

		xasprintf(&csock, "%s/amused-%d", tmpdir, getuid());
	}

	if (argc > 0)
		debug = 0;

	ctl(argc, argv);
}

void
spawn_daemon(void)
{
	start_child(PROC_MAIN, -1);
}

void
imsg_event_add(struct imsgev *iev)
{
	iev->events = EV_READ;
	if (imsgbuf_queuelen(&iev->imsgbuf))
		iev->events |= EV_WRITE;

	ev_add(iev->imsgbuf.fd, iev->events, iev->handler, iev);
}

int
imsg_compose_event(struct imsgev *iev, uint16_t type, uint32_t peerid,
    pid_t pid, int fd, const void *data, uint16_t datalen)
{
	int ret;

	if ((ret = imsg_compose(&iev->imsgbuf, type, peerid, pid, fd, data,
	    datalen)) != -1)
		imsg_event_add(iev);

	return ret;
}

int
main_send_player(uint16_t type, int fd, const void *data, size_t len)
{
	return imsg_compose_event(iev_player, type, 0, 0, fd, data, len);
}

int
main_play_song(const char *path)
{
	struct stat sb;
	int fd;

	if ((fd = open(path, O_RDONLY)) == -1) {
		log_warn("open %s", path);
		return 0;
	}

	if (fstat(fd, &sb) == -1) {
		log_warn("failed to stat %s", path);
		close(fd);
		return 0;
	}

	if (!S_ISREG(sb.st_mode)) {
		log_info("skipping non-regular file: %s", path);
		close(fd);
		return 0;
	}

	play_state = STATE_PLAYING;
	main_send_player(IMSG_PLAY, fd, NULL, 0);
	return 1;
}

void
main_playlist_jump(struct imsgev *iev, struct imsg *imsg)
{
	char arg[PATH_MAX];
	const char *song;

	if (imsg_get_data(imsg, arg, sizeof(arg)) == -1) {
		main_senderr(iev, "wrong size");
		return;
	}

	if (arg[sizeof(arg)-1] != '\0') {
		main_senderr(iev, "data corrupted");
		return;
	}

	song = playlist_jump(arg);
	if (song == NULL) {
		main_senderr(iev, "not found");
		return;
	}

	control_notify(IMSG_CTL_JUMP);

	main_send_player(IMSG_STOP, -1, NULL, 0);
	if (!main_play_song(song)) {
		main_senderr(iev, "can't play");
		playlist_dropcurrent();
		main_playlist_advance();
		return;
	}

	main_send_status(iev);
}

void
main_playlist_resume(void)
{
	const char *song;

	if ((song = current_song) == NULL)
		song = playlist_advance();

	for (; song != NULL; song = playlist_advance()) {
		if (main_play_song(song))
			return;

		playlist_dropcurrent();
	}
}

void
main_playlist_advance(void)
{
	const char *song;

	for (;;) {
		song = playlist_advance();
		if (song == NULL)
			return;

		if (main_play_song(song))
			break;

		playlist_dropcurrent();
	}
}

void
main_playlist_previous(void)
{
	const char *song;

	for (;;) {
		song = playlist_previous();
		if (song == NULL)
			return;

		if (main_play_song(song))
			break;

		playlist_dropcurrent();
	}
}

void
main_senderr(struct imsgev *iev, const char *msg)
{
	imsg_compose_event(iev, IMSG_CTL_ERR, 0, 0, -1,
	    msg, strlen(msg)+1);
}

void
main_enqueue(int tx, struct playlist *px, struct imsgev *iev,
    struct imsg *imsg)
{
	char path[PATH_MAX];
	const char *err = NULL;

	if (imsg_get_data(imsg, path, sizeof(path)) == -1) {
		err = "data size mismatch";
		goto err;
	}

	if (path[sizeof(path)-1] != '\0') {
		err = "malformed data";
		goto err;
	}

	if (tx)
		playlist_push(px, path);
	else
		playlist_enqueue(path);
	imsg_compose_event(iev, IMSG_CTL_ADD, 0, 0, -1, path, sizeof(path));
	return;
err:
	main_senderr(iev, err);
}

void
main_send_playlist(struct imsgev *iev)
{
	struct player_status s;
	size_t i;

	for (i = 0; i < playlist.len; ++i) {
		memset(&s, 0, sizeof(s));
		strlcpy(s.path, playlist.songs[i], sizeof(s.path));
		s.status = play_off == i ? STATE_PLAYING : STATE_STOPPED;
		imsg_compose_event(iev, IMSG_CTL_SHOW, 0, 0, -1, &s,
		    sizeof(s));
	}

	imsg_compose_event(iev, IMSG_CTL_SHOW, 0, 0, -1, NULL, 0);
}

void
main_send_status(struct imsgev *iev)
{
	struct player_status s;

	memset(&s, 0, sizeof(s));

	if (current_song != NULL)
		strlcpy(s.path, current_song, sizeof(s.path));
	s.status = play_state;
	s.position = current_status.position;
	s.duration = current_status.duration;
	s.mode.repeat_all = repeat_all;
	s.mode.repeat_one = repeat_one;
	s.mode.consume = consume;
	s.info.bits = current_status.info.bits;
	s.info.rate = current_status.info.rate;
	s.info.chan = current_status.info.chan;

	imsg_compose_event(iev, IMSG_CTL_STATUS, 0, 0, -1, &s, sizeof(s));
}

void
main_seek(struct player_seek *s)
{
	switch (play_state) {
	case STATE_STOPPED:
		main_playlist_resume();
		break;
	case STATE_PLAYING:
		break;
	case STATE_PAUSED:
		play_state = STATE_PLAYING;
		break;
	}

	main_send_player(IMSG_CTL_SEEK, -1, s, sizeof(*s));
}
